diff -ruN libtorrent-rasterbar-1.1.8.orig/include/libtorrent/Makefile.am libtorrent-rasterbar-1.1.8/include/libtorrent/Makefile.am
--- libtorrent-rasterbar-1.1.8.orig/include/libtorrent/Makefile.am	2018-06-25 16:04:14.000000000 +0700
+++ libtorrent-rasterbar-1.1.8/include/libtorrent/Makefile.am	2018-06-25 19:22:54.553872074 +0700
@@ -154,11 +154,6 @@
   web_peer_connection.hpp      \
   xml_parse.hpp                \
   \
-  tommath.h                         \
-  tommath_class.h                   \
-  tommath_superclass.h              \
-  tommath_private.h                 \
-  \
   aux_/alert_manager_variadic_emplace.hpp \
   aux_/allocating_handler.hpp       \
   aux_/cpuid.hpp                    \
diff -ruN libtorrent-rasterbar-1.1.8.orig/include/libtorrent/Makefile.in libtorrent-rasterbar-1.1.8/include/libtorrent/Makefile.in
--- libtorrent-rasterbar-1.1.8.orig/include/libtorrent/Makefile.in	2018-06-25 17:30:15.000000000 +0700
+++ libtorrent-rasterbar-1.1.8/include/libtorrent/Makefile.in	2018-06-25 19:22:54.554872078 +0700
@@ -495,11 +495,6 @@
   web_peer_connection.hpp      \
   xml_parse.hpp                \
   \
-  tommath.h                         \
-  tommath_class.h                   \
-  tommath_superclass.h              \
-  tommath_private.h                 \
-  \
   aux_/alert_manager_variadic_emplace.hpp \
   aux_/allocating_handler.hpp       \
   aux_/cpuid.hpp                    \
diff -ruN libtorrent-rasterbar-1.1.8.orig/src/Makefile.am libtorrent-rasterbar-1.1.8/src/Makefile.am
--- libtorrent-rasterbar-1.1.8.orig/src/Makefile.am	2018-06-25 16:04:14.000000000 +0700
+++ libtorrent-rasterbar-1.1.8/src/Makefile.am	2018-06-25 19:22:54.554872078 +0700
@@ -92,7 +92,6 @@
   magnet_uri.cpp                  \
   merkle.cpp                      \
   metadata_transfer.cpp           \
-  mpi.cpp                         \
   natpmp.cpp                      \
   parse_url.cpp                   \
   part_file.cpp                   \
@@ -158,7 +157,7 @@
   $(BUILTIN_CRYPTO_SOURCES)
 
 libtorrent_rasterbar_la_LDFLAGS = -version-info $(INTERFACE_VERSION_INFO)
-libtorrent_rasterbar_la_LIBADD = @BOOST_SYSTEM_LIB@ @OPENSSL_LIBS@
+libtorrent_rasterbar_la_LIBADD = @BOOST_SYSTEM_LIB@ @OPENSSL_LIBS@ -ltommath
 
 AM_CPPFLAGS = -DTORRENT_BUILDING_LIBRARY -I$(top_srcdir)/include -I$(top_srcdir)/ed25519/src @DEBUGFLAGS@ @OPENSSL_INCLUDES@
 AM_CFLAGS = -I$(top_srcdir)/ed25519/src -std=c99
diff -ruN libtorrent-rasterbar-1.1.8.orig/src/Makefile.in libtorrent-rasterbar-1.1.8/src/Makefile.in
--- libtorrent-rasterbar-1.1.8.orig/src/Makefile.in	2018-06-25 17:30:16.000000000 +0700
+++ libtorrent-rasterbar-1.1.8/src/Makefile.in	2018-06-25 19:22:25.329732560 +0700
@@ -156,7 +156,7 @@
 	http_tracker_connection.cpp i2p_stream.cpp identify_client.cpp \
 	instantiate_connection.cpp ip_filter.cpp ip_voter.cpp \
 	lazy_bdecode.cpp lsd.cpp lt_trackers.cpp magnet_uri.cpp \
-	merkle.cpp metadata_transfer.cpp mpi.cpp natpmp.cpp \
+	merkle.cpp metadata_transfer.cpp natpmp.cpp \
 	parse_url.cpp part_file.cpp pe_crypto.cpp \
 	performance_counters.cpp peer_connection.cpp \
 	peer_connection_handle.cpp peer_class.cpp peer_class_set.cpp \
@@ -223,7 +223,7 @@
 	http_tracker_connection.lo i2p_stream.lo identify_client.lo \
 	instantiate_connection.lo ip_filter.lo ip_voter.lo \
 	lazy_bdecode.lo lsd.lo lt_trackers.lo magnet_uri.lo merkle.lo \
-	metadata_transfer.lo mpi.lo natpmp.lo parse_url.lo \
+	metadata_transfer.lo natpmp.lo parse_url.lo \
 	part_file.lo pe_crypto.lo performance_counters.lo \
 	peer_connection.lo peer_connection_handle.lo peer_class.lo \
 	peer_class_set.lo piece_picker.lo platform_util.lo \
@@ -649,7 +649,6 @@
   magnet_uri.cpp                  \
   merkle.cpp                      \
   metadata_transfer.cpp           \
-  mpi.cpp                         \
   natpmp.cpp                      \
   parse_url.cpp                   \
   part_file.cpp                   \
@@ -715,7 +714,7 @@
   $(BUILTIN_CRYPTO_SOURCES)
 
 libtorrent_rasterbar_la_LDFLAGS = -version-info $(INTERFACE_VERSION_INFO)
-libtorrent_rasterbar_la_LIBADD = @BOOST_SYSTEM_LIB@ @OPENSSL_LIBS@
+libtorrent_rasterbar_la_LIBADD = @BOOST_SYSTEM_LIB@ @OPENSSL_LIBS@ -ltommath
 AM_CPPFLAGS = -DTORRENT_BUILDING_LIBRARY -I$(top_srcdir)/include -I$(top_srcdir)/ed25519/src @DEBUGFLAGS@ @OPENSSL_INCLUDES@
 AM_CFLAGS = -I$(top_srcdir)/ed25519/src -std=c99
 AM_LDFLAGS = @OPENSSL_LDFLAGS@
@@ -844,7 +844,6 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/magnet_uri.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/merkle.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/metadata_transfer.Plo@am__quote@
-@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/mpi.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/natpmp.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/packet_buffer.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/parse_url.Plo@am__quote@
diff -ruN libtorrent-rasterbar-1.1.8.orig/src/pe_crypto.cpp libtorrent-rasterbar-1.1.8/src/pe_crypto.cpp
--- libtorrent-rasterbar-1.1.8.orig/src/pe_crypto.cpp	2018-06-25 16:04:14.000000000 +0700
+++ libtorrent-rasterbar-1.1.8/src/pe_crypto.cpp	2018-06-25 19:22:54.555872083 +0700
@@ -38,7 +38,7 @@
 #include <algorithm>
 
 extern "C" {
-#include "libtorrent/tommath.h"
+#include <tommath.h>
 }
 
 #include "libtorrent/aux_/disable_warnings_pop.hpp"
diff -ruN libtorrent-rasterbar-1.1.8.orig/test/test_pe_crypto.cpp libtorrent-rasterbar-1.1.8/test/test_pe_crypto.cpp
--- libtorrent-rasterbar-1.1.8.orig/test/test_pe_crypto.cpp	2018-06-25 16:04:14.000000000 +0700
+++ libtorrent-rasterbar-1.1.8/test/test_pe_crypto.cpp	2018-06-25 19:22:54.566872136 +0700
@@ -42,7 +42,7 @@
 #include "test.hpp"
 
 extern "C" {
-#include "libtorrent/tommath.h"
+#include <tommath.h>
 }
 
 #if !defined(TORRENT_DISABLE_ENCRYPTION) && !defined(TORRENT_DISABLE_EXTENSIONS)
